crates/libcontainer/src/container/container_delete.rs

```diff
@@ -27,14 +27,48 @@ impl Container {
     pub fn delete(&mut self, force: bool) -> Result<()> {
         self.refresh_status()
             .context("failed to refresh container status")?;
-        if self.can_kill() && force {
-            self.do_kill(signal::Signal::SIGKILL, true)?;
-            self.set_status(ContainerStatus::Stopped).save()?;
-        }
+
         log::debug!("container status: {:?}", self.status());
-        if self.can_delete() {
-            if self.root.exists() {
-                if let Ok(config) = YoukiConfig::load(&self.root) {
+
+        // Check if container is allowed to be deleted based on container status.
+        match self.status() {
+            ContainerStatus::Stopped => {}
+            ContainerStatus::Created => {
+                // Here, we differ from the OCI spec, but matches the same
+                // behavior as `runc` and `crun`. The OCI spec does not allow
+                // deletion of status `created` without `force` flag. But both
+                // `runc` and `crun` allows deleting `created`. Therefore we
+                // decided to follow `runc` and `crun`. In the case where
+                // container is in `created` status, we first kill the container
+                // init process which is waiting on start signal. Since only a
+                // single container process (the init process) is created, we do
+                // not need to send signals to all in this case.
+                self.do_kill(signal::Signal::SIGKILL, false)?;
+                self.set_status(ContainerStatus::Stopped).save()?;
+            }
+            ContainerStatus::Creating | ContainerStatus::Running | ContainerStatus::Paused => {
```

Comment:
Should the `Creating` state also handled along with `Created` ? If I'm correct, then even at the time of `Creating` state, there should be only one process in the container's cgroup. If that is correct, then we don't get any advantage / have any reason to use kill all with force. I checked the spec, and it specifies that states other than `stopped` should generate error, but we are going according to runc anyways.

Please correct me If my understanding was wrong.

Reply:
Technically 2 process, container init and intermediate process before it exits. With that being said, it is possible I am over optimizing in this case. The goal here is to kill all the process in the container cgroup, so I should possibly just call kill all here for either cases.

Reply:
Hey, If there are two processes, then this logic makes perfect sense.
Also, I don't think we should merge `Created` with rest 3 (if that was what you meant) , because as per runc and crun, we need to allow deleting `Created` without force flag, however need force flag for others.

If there are going to be (even technically) two processes, in `Creating` state, it is better to handle it by killing all in cgroup, just to be safe.